File: backend/ruby_foundations/exercises/small_problems_101/easy_7/lettercase_counter.rb
def letter_case_count(string)
  stat = {lowercase: 0, uppercase: 0, neither: 0}
  string.chars.each do |letter|
    case letter
    when /[a-z]/
      stat[:lowercase] += 1
    when /[A-Z]/
      stat[:uppercase] += 1
    else
      stat[:neither] += 1
    end
  end
  return stat
end

puts letter_case_count('abCdef 123') == { lowercase: 5, uppercase: 1, neither: 4 }
puts letter_case_count('AbCd +Ef') == { lowercase: 3, uppercase: 3, neither: 2 }
puts letter_case_count('123') == { lowercase: 0, uppercase: 0, neither: 3 }
puts letter_case_count('') == { lowercase: 0, uppercase: 0, neither: 0 }

File: frontend/javascript/exercises/string_and_text_processing/lettercase_counter.js
function isLowerCase(char) {
  return char >= 'a' && char <= 'z'
}

function isUpperCase(char) {
  return char >= 'A' && char <= 'Z'
}

function isNeither(char) {
  return !isLowerCase(char) && !isUpperCase(char);
}

function letterCaseCount(string) {
  var counter = {lowercase: 0, uppercase: 0, neither: 0};
  counter.uppercase = string.split('').filter(isUpperCase).length;
  counter.lowercase = string.split('').filter(isLowerCase).length;
  counter.neither = string.split('').filter(isNeither).length;
  return counter;
}

// original solution:
// function letterCaseCount(string) {
//   var counter = {lowercase: 0, uppercase: 0, neither: 0};
//   for (var i = 0; i < string.length; i++) {
//     if (isUpperCase(string[i])) {
//       counter.uppercase += 1;
//     } else if (isLowerCase(string[i])) {
//       counter.lowercase += 1;
//     } else {
//       counter.neither += 1;
//     }
//   }
//   return counter;
// }

console.log(letterCaseCount('abCdef 123')); // { lowercase: 5, uppercase: 1, neither: 4 }
console.log(letterCaseCount('AbCd +Ef'));   // { lowercase: 3, uppercase: 3, neither: 2 }
console.log(letterCaseCount('123'));        // { lowercase: 0, uppercase: 0, neither: 3 }
console.log(letterCaseCount(''));           // { lowercase: 0, uppercase: 0, neither: 0 }